flink-tests/src/test/java/org/apache/flink/test/checkpointing/ImmediateCheckpointingAfterAllTasksFinishedITCase.java

```diff
@@ -0,0 +1,202 @@
+import org.apache.flink.runtime.minicluster.MiniCluster;
+import org.apache.flink.runtime.minicluster.MiniClusterConfiguration;
+import org.apache.flink.runtime.state.StateSnapshotContext;
+import org.apache.flink.runtime.testutils.CommonTestUtils;
+import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
+import org.apache.flink.streaming.api.functions.source.RichSourceFunction;
+import org.apache.flink.streaming.api.graph.StreamGraph;
+import org.apache.flink.streaming.api.operators.AbstractStreamOperator;
+import org.apache.flink.streaming.api.operators.OneInputStreamOperator;
+import org.apache.flink.streaming.runtime.streamrecord.StreamRecord;
+import org.apache.flink.test.util.AbstractTestBase;
+
+import org.junit.Assert;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.io.TempDir;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.ValueSource;
+
+/**
+ * Tests an immediate checkpoint should be triggered right after all tasks reached the end of data.
+ */
+public class ImmediateCheckpointingAfterAllTasksFinishedITCase extends AbstractTestBase {
+    private static final int SMALL_SOURCE_NUM_RECORDS = 20;
+    private static final int BIG_SOURCE_NUM_RECORDS = 100;
+
+    private StreamExecutionEnvironment env;
+
+    @TempDir private java.nio.file.Path tmpDir;
+
+    @BeforeEach
+    public void setUp() {
+        env = StreamExecutionEnvironment.getExecutionEnvironment();
+        env.setRestartStrategy(RestartStrategies.noRestart());
+        env.enableCheckpointing(Long.MAX_VALUE - 1);
+    }
+
+    @ParameterizedTest
+    @ValueSource(ints = {1, 4})
+    public void testImmediateCheckpointing(int parallelism) throws Exception {
+        env.setParallelism(parallelism);
+        StreamGraph streamGraph = getStreamGraph(env);
+        env.execute(streamGraph);
+    }
+
+    @ParameterizedTest
+    @ValueSource(ints = {1, 4})
+    public void testRestoreAfterSomeTasksFinished(int parallelism) throws Exception {
+        env.setParallelism(parallelism);
+
+        final MiniClusterConfiguration cfg =
+                new MiniClusterConfiguration.Builder()
+                        .withRandomPorts()
+                        .setNumTaskManagers(1)
+                        .setNumSlotsPerTaskManager(4)
+                        .build();
+        try (MiniCluster miniCluster = new MiniCluster(cfg)) {
+            miniCluster.start();
+
+            env.enableCheckpointing(100);
+            JobGraph jobGraph = getStreamGraph(env).getJobGraph();
+            miniCluster.submitJob(jobGraph).get();
+
+            CommonTestUtils.waitForSubtasksToFinish(
+                    miniCluster,
+                    jobGraph.getJobID(),
+                    findVertexByName(jobGraph, "Source: smallSource").getID(),
+                    false);
+
+            String savepointPath =
+                    miniCluster
+                            .triggerSavepoint(
```

Comment:
It is almost impossible that all tasks have stopped since the two pipelines of the Job have a different time span of generating source data(200ms vs 1000ms) and the `CommonTestUtils.waitForSubtasksToFinish` would return as long as the short one finishes. 
Maybe it would cost 800ms to trigger the savepoint in extreme cases like the JVM is full GC, but I think it is enough for a regression test that is run every day and we don't have to increase the complexity of the code for it. How do you think?

Reply:
While it is likely that we can eventually catch error given that this test runs every day, it is still preferred to avoid having flaky test if possible.

Also note that It is preferred not to use sleep in CI. Though it might seem OK to sleep for 1 sec in this test, the overall CI time can be several times higher than necessary if every test does this.

Overall, I feel the benefit of reducing test execution time and making the test less flaky is probably worth the one-time development cost. Note that many existing tests use synchronization primitives (e.g. CountDownLatch) instead of sleep for cases like this. Maybe you can checkout the code https://github.com/apache/flink/blob/master/flink-tests/src/test/java/org/apache/flink/test/checkpointing/RescalingITCase.java#L322 for example.